tests/integration/test_core.py

```diff
@@ -53,6 +55,33 @@ async def test_hello_kubecon_cache_header(ingress_ip: str):
     assert "content-cache-k8s" in response.headers["X-Cache-Status"]
 
 
+@pytest.mark.asyncio
+@pytest.mark.abort_on_fail
+async def test_service_reachable(service_ip: str):
+    """
+    arrange: given charm has been built, deployed and related to a dependent application
+    act: when the dependent application is queried via the service
+    assert: then the response is HTTP 200 OK.
+    """
+    response = requests.get(f"http://{service_ip}", timeout=5)
+
+    assert response.status_code == 200
+
+
+async def test_report_visits_by_ip(app: Application):
```

Comment:
This test seems quite light, what about cases where multiple IPS have visited? Are we sure our code works in that case? What bout if an IP has not visited in the last 20 minutes? What if it visited just before 20 minutes ago? What if it visited exactly 20 minutes ago? What if there are 2 or more IPs that visited the same number of times in the last 20 minutes? And so on

This is why it might potentially be easier to write a Python function that goes through a string to calculate the result because you can easily test it with input output parametrised tests.